### DISCUSSION TITLE
Dynamic diagnostics and modified indicator; obviates my need for a status line!

### DISCUSSION THREAD

**Discussion by raj-magesh** [b0o/incline.nvim#31] [Showcase]
![2022-10-26T0114](https://user-images.githubusercontent.com/28287009/197940592-01d4f25e-6b39-48a9-95ca-a4b44564dba8.png)
![2022-10-26T0111](https://user-images.githubusercontent.com/28287009/197940326-ca91bc3e-fca4-4053-a53f-80c3b265114b.png)

Building on this [discussion](#26) that gives an example of how diagnostic info can be added to `incline`! This is my first time coding in `lua` (with any control flow, at least), so if you spot any places to improve, please point them out!

## Configuration

```lua
local function get_diagnostic_label(props)
    local icons = {
        Error = "",
        Warn = "",
        Info = "",
        Hint = "",
    }

    local label = {}
    for severity, icon in pairs(icons) do
        local n = #vim.diagnostic.get(props.buf, {severity = vim.diagnostic.severity[string.upper(severity)]})
        if n > 0 then
            local fg = "#" .. string.format("%06x", vim.api.nvim_get_hl_by_name("DiagnosticSign" .. severity, true)["foreground"])
            table.insert(label, {icon .. " " .. n .. " ", guifg = fg})
        end
    end
    return label
end

require("incline").setup({
  debounce_threshold = { falling = 500, rising = 250 },
  render = function(props)
    local bufname = vim.api.nvim_buf_get_name(props.buf)
    local filename = vim.fn.fnamemodify(bufname, ":t")
    local diagnostics = get_diagnostic_label(props)
    local modified = vim.api.nvim_buf_get_option(props.buf, "modified") and "bold,italic" or "None"
    local filetype_icon, color = require("nvim-web-devicons").get_icon_color(filename)

    local buffer = {
        { filetype_icon, guifg = color },
        { " " },
        { filename, gui = modified },
    }

    if #diagnostics > 0 then
        table.insert(diagnostics, { "| ", guifg = "grey" })
    end
    for _, buffer_ in ipairs(buffer) do
        table.insert(diagnostics, buffer_)
    end
    return diagnostics
  end,
})
```

## Metadata

colorschemes: [Gruvbox](https://github.com/ellisonleao/gruvbox.nvim), [Catppuccin (macchiato)](https://github.com/catppuccin/nvim)
font: [Meslo Nerd Font](https://github.com/ryanoasis/nerd-fonts/releases/download/v2.2.2/Meslo.zip)

**Comment by b0o** [comment]
This looks great, I think I might steal the idea for my config!

One suggestion, I think you can simplify the code a bit:


```lua
local function get_diagnostic_label(props)
  local icons = {
    Error = '',
    Warn = '',
    Info = '',
    Hint = '',
  }

  local label = {}
  for severity, icon in pairs(icons) do
    local n = #vim.diagnostic.get(props.buf, { severity = vim.diagnostic.severity[string.upper(severity)] })
    if n > 0 then
      table.insert(label, { icon .. ' ' .. n .. ' ', group = 'DiagnosticSign' .. severity })
    end
  end
  return label
end
```

Unless you really only want the foreground color, but I tested it and it seems to look the same in my theme at least.

**Reply by raj-magesh** [reply]
TIL about the `group` key, thanks!

Do you by any chance know how I could access the icon directly from my definition of the diagnostic sign elsewhere? I have the following in my `init.lua`:

```lua
    local signs = {
        Error = " ",
        Warn = " ",
        Info = " ",
        Hint = " ",
    }

    for type_, icon in pairs(signs) do
        local hl = "DiagnosticSign" .. type_
        vim.fn.sign_define(hl, { text = icon, texthl = hl })
    end
```

It's repetitive to have the definitions for the icons in two places. I skimmed the neovim docs but couldn't find an obvious way to access the text icon of an already defined sign.

**Reply by b0o** [reply]
It seems there's a `sign_getdefined` function: https://neovim.io/doc/user/sign.html#sign_getdefined()

Or you could create a separate lua file `signs.lua` (preferably, place it in a unique subfolder to namespace it, [I personally use `lua/user`](https://github.com/b0o/nvim-conf/tree/main/lua/user)):

```lua
local signs = {
    Error = " ",
    Warn = " ",
    Info = " ",
    Hint = " ",
}

return signs
```

And then require the file wherever you need to use the signs:

```lua
local signs = require'user.signs' 
```

**Comment by vollowx** [comment]
Wondering how can you make such a LSP diagnostics like this

```
aaa aaa aaa - Hint
|    |- Warning
|- Error
```

**Reply by raj-magesh** [reply]
It's the lsp_lines plugin by WhyNotHugo!
